superset-frontend/src/SqlLab/components/SaveDatasetModal/index.tsx

```diff
@@ -74,96 +79,290 @@ const Styles = styled.div`
+        columns: selectedColumns,
+      })
+      .then((data: { table_id: number }) => {
+        exploreChart({
+          datasource: `${data.table_id}__table`,
+          metrics: [],
+          groupby: [],
+          time_range: 'No filter',
+          viz_type: 'table',
+          all_columns: selectedColumns.map(c => c.name),
+          row_limit: 1000,
+        });
+      })
+      .catch(() => {
+        actions.addDangerToast(t('An error occurred saving dataset'));
+      });
+
+    setNewSaveDatasetName(getDefaultDatasetName());
+    onHide();
+  };
+
+  const handleSaveDatasetModalSearch = async (searchText: string) => {
+    const userDatasetsOwned = await getUserDatasets(searchText);
+    setUserDatasetOptions(userDatasetsOwned);
+  };
+
+  const handleOverwriteDatasetOption = (
+    _data: string,
+    option: Record<string, any>,
+  ) => setDatasetToOverwrite(option);
+
+  const handleDatasetNameChange = (e: React.FormEvent<HTMLInputElement>) => {
+    // @ts-expect-error
+    setNewSaveDatasetName(e.target.value);
+  };
+
+  const handleSaveDatasetRadioBtnState = (e: RadioChangeEvent) => {
+    setSaveDatasetRadioBtnState(Number(e.target.value));
+  };
+
+  const handleOverwriteCancel = () => {
+    setShouldOverwriteDataset(false);
+    setDatasetToOverwrite({});
+  };
+
+  const disableSaveAndExploreBtn =
+    (saveDatasetRadioBtnState === DatasetRadioState.SAVE_NEW &&
+      newSaveDatasetName.length === 0) ||
+    (saveDatasetRadioBtnState === DatasetRadioState.OVERWRITE_DATASET &&
+      Object.keys(datasetToOverwrite).length === 0 &&
+      saveModalAutocompleteValue.length === 0);
+
+  const filterAutocompleteOption = (
+    inputValue: string,
+    option: { value: string; datasetId: number },
+  ) => option.value.toLowerCase().includes(inputValue.toLowerCase());
+
+  return (
+    <StyledModal
+      show={visible}
+      title={t('Save or Overwrite Dataset')}
+      onHide={onHide}
+      footer={
+        <>
+          {!shouldOverwriteDataset && (
             <Button
-              className="md"
-              buttonStyle="primary"
-              onClick={handleOverwriteDataset}
               disabled={disableSaveAndExploreBtn}
+              buttonStyle="primary"
+              onClick={handleSaveInDataset}
             >
-              {t('Overwrite & Explore')}
+              {buttonTextOnSave}
             </Button>
-          </>
+          )}
+          {shouldOverwriteDataset && (
+            <>
+              <Button onClick={handleOverwriteCancel}>Back</Button>
+              <Button
+                className="md"
+                buttonStyle="primary"
+                onClick={handleOverwriteDataset}
+                disabled={disableSaveAndExploreBtn}
+              >
+                {buttonTextOnOverwrite}
+              </Button>
+            </>
+          )}
+        </>
+      }
+    >
+      <Styles>
+        {!shouldOverwriteDataset && (
+          <div className="smd-body">
```

Comment:
@lyndsiWilliams do you mind cleaning up the old code here too while we're at it? Really small nit, but I think this is supposed to be `sdm` for save dataset modal? :)